In [1]:
import pandas as pd
import numpy as np 
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler

In [93]:
df=pd.read_csv('/Users/bryan/code/VonRiecken/Power-Predict/power_predict/data/merged_dataset.csv')
df

Unnamed: 0  Month_year         Country                     Balance  \
0              0  2010-04-01       Australia  Net Electricity Production   
1              1  2010-04-01         Austria  Net Electricity Production   
2              2  2010-04-01         Belgium  Net Electricity Production   
3              3  2010-04-01          Canada  Net Electricity Production   
4              4  2010-04-01           Chile  Net Electricity Production   
...          ...         ...             ...                         ...   
6713        6713  2022-09-01           Spain  Net Electricity Production   
6714        6714  2022-09-01          Sweden  Net Electricity Production   
6715        6715  2022-09-01     Switzerland  Net Electricity Production   
6716        6716  2022-09-01  United Kingdom  Net Electricity Production   
6717        6717  2022-09-01   United States  Net Electricity Production   

      Combustible Renewables       Hydro  Other Renewables       Solar  \
0                   216.2870   1044.4060            0.0000     26.8110   
1                   350.3830   2504.1300            0.0000      9.9650   
2                   383.1770    141.8980            0.0000     80.4190   
3                   712.9880  28243.7380            2.4400     24.1840   
4                   152.8260   1740.4160            0.0000      0.0000   
...                      ...         ...               ...         ...   
6713                478.4796   1459.6225            1.5000   3213.8391   
6714                819.7182   5992.2339            0.0000    101.3747   
6715                173.8267   2753.3512            0.0000    337.7122   
6716               3122.6177    367.4721            0.4559   1385.7783   
6717               4322.0950  18246.4312            0.0000  17437.4229   

      Total Renewables (Hydro, Geo, Solar, Wind, Other)        Wind  \
0                                             1638.0980    350.5110   
1                                             2995.6960    131.1070   
2                                              702.5090     97.0150   
3                                            29709.3010    725.9510   
4                                             1913.6010     20.3590   
...                                                 ...         ...   
6713                                          9351.9487   4198.5075   
6714                                          8900.2441   1986.9173   
6715                                          3276.0098     11.1196   
6716                                         10397.5687   5521.2446   
6717                                         68452.5391  27065.9698   

      value_CDD_18  value_CDD_21  value_Global_Horizontal_Irrandiance  \
0         48.13000     12.620000                             720994.0   
1          0.05449      0.000000                             736161.0   
2          0.99160      0.000000                             723195.0   
3          0.06883      0.000023                             679927.0   
4          1.41100      0.040190                             578355.0   
...            ...           ...                                  ...   
6713     112.80000     48.410000                             748831.0   
6714       0.00000      0.000000                             314577.0   
6715       3.99700      0.000000                             565099.0   
6716       2.85300      0.000000                             422664.0   
6717     137.80000     74.390000                             697004.0   

      value_HDD_16  value_HDD_18  value_Heat_index  value_Relative_Humidty  \
0           17.470         37.08            18.410                   57.37   
1          228.900        287.80             8.410                   69.85   
2          195.800        253.70             9.577                   67.90   
3          246.700        306.20             7.797                   70.85   
4          113.000        168.30            12.440                   66.58   
...   

In [94]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6718 entries, 0 to 6717
Data columns (total 20 columns):
 #   Column                                             Non-Null Count  Dtype  
---  ------                                             --------------  -----  
 0   Unnamed: 0                                         6718 non-null   int64  
 1   Month_year                                         6718 non-null   object 
 2   Country                                            6718 non-null   object 
 3   Balance                                            6718 non-null   object 
 4   Combustible Renewables                             6718 non-null   float64
 5   Hydro                                              6718 non-null   float64
 6   Other Renewables                                   6718 non-null   float64
 7   Solar                                              6718 non-null   float64
 8   Total Renewables (Hydro, Geo, Solar, Wind, Other)  6718 non-null   float64
 9   Wind    

In [26]:
print(df.columns)


Index(['Unnamed: 0', 'Month_year', 'Country', 'Balance',
       'Combustible Renewables', 'Hydro', 'Other Renewables', 'Solar',
       'Total Renewables (Hydro, Geo, Solar, Wind, Other)', 'Wind',
       'value_CDD_18', 'value_CDD_21', 'value_Global_Horizontal_Irrandiance',
       'value_HDD_16', 'value_HDD_18', 'value_Heat_index',
       'value_Relative_Humidty', 'value_Temperature',
       'value_Total_Precipitation', 'total target (wind, solar, hydro)'],
      dtype='object')


### One Hot Encoder Scaled model

In [85]:
numerical_columns = X.select_dtypes(include=['float64', 'int64']).columns

# StandardScaler object
scaler = StandardScaler()

# Scaled numerical features
X[numerical_columns] = scaler.fit_transform(X[numerical_columns])
X[numerical_columns]

Unnamed: 0  Combustible Renewables     Hydro  Other Renewables  \
0      -1.731793               -0.371839 -0.288311         -0.219984   
1      -1.731277               -0.284767 -0.186769         -0.219984   
2      -1.730762               -0.263473 -0.351091         -0.219984   
3      -1.730246               -0.049320  1.603731          0.079841   
4      -1.729730               -0.413045 -0.239895         -0.219984   
...          ...                     ...       ...               ...   
6713    1.729730               -0.201591 -0.259427         -0.035665   
6714    1.730246                0.019982  0.055871         -0.219984   
6715    1.730762               -0.399409 -0.169433         -0.219984   
6716    1.731277                1.515304 -0.335400         -0.163964   
6717    1.731793                2.294151  0.908298         -0.219984   

         Solar  Total Renewables (Hydro, Geo, Solar, Wind, Other)      Wind  \
0    -0.279923                                          -0.316003 -0.268083   
1    -0.285258                                          -0.257861 -0.302890   
2    -0.262948                                          -0.356072 -0.308299   
3    -0.280755                                           0.886205 -0.208521   
4    -0.288413                                          -0.304204 -0.320460   
...        ...                                                ...       ...   
6713  0.729301                                           0.014359  0.342387   
6714 -0.256312                                          -0.004987 -0.008473   
6715 -0.181471                                          -0.245856 -0.321926   
6716  0.150416                                           0.059139  0.552234   
6717  5.233428                                           2.545465  3.970220   

      value_CDD_18  value_CDD_21  value_Global_Horizontal_Irrandiance  \
0         0.139958     -0.151031                             0.523027   
1        -0.559862     -0.445247                             0.573405   
2        -0.546221     -0.445247                             0.530337   
3        -0.559654     -0.445247                             0.386620   
4        -0.540116     -0.444310                             0.049241   
...            ...           ...                                  ...   
6713      1.081340      0.683360                             0.615489   
6714     -0.560656     -0.445247                            -0.826916   
6715     -0.502473     -0.445247                             0.005211   
6716     -0.519125     -0.445247                            -0.467897   
6717      1.445258      1.289045                             0.443342   

      value_HDD_16  value_HDD_18  value_Heat_index  value_Relative_Humidty  \
0        -0.874051     -0.913969          0.754759               -1.661187   
1         0.311842      0.371240         -0.473498               -0.406227   
2         0.126186      0.196441         -0.330161               -0.602314   
3         0.411680      0.465560         -0.548791               -0.305669   
4        -0.338232     -0.241326          0.021489               -0.735051   
...            ...           ...               ...                     ...   
6713     -0.940186     -1.034073          1.117095               -1.372586   
6714     -0.203618     -0.094207         -0.102565                0.653651   
6715     -0.358424     -0.296175          0.075533                0.545049   
6716     -0.603702     -0.523260          0.252402                0.666723   
6717     -0.914099     -0.998755          1.220269               -1.192588   

      value_Temperature  value_Total_Precipitation  
0              1.350500                  -0.775001  
1             -0.518854                  -0.144525  
2             -0.226562                  -0.961121  
3             -1.479081                  -0.778387  
4             -0.186817                   0.516427  
...                 ...                        ...  
6713   

### Logistical Regression Scaled

In [28]:
from sklearn.metrics import mean_squared_error, r2_score

numerical_columns = X.select_dtypes(include=['float64', 'int64']).columns

# StandardScaler object
scaler = StandardScaler()

# Scaled numerical features
X[numerical_columns] = scaler.fit_transform(X[numerical_columns])

# Split data train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# linear regression model
model = LinearRegression()

# Fit 
model.fit(X_train, y_train)

# predictions on the testing data
y_pred = model.predict(X_test)

# Mean Absolute Error
mae = mean_absolute_error(y_test, y_pred)

# Mean Squared Error
mse = mean_squared_error(y_test, y_pred)

# Coefficient of Determination (COD)
cod = r2_score(y_test, y_pred)

r2 = r2_score(y_test, y_pred)

print(f'Mean Absolute Error: {mae}')
print(f'Mean Squared Error: {mse}')
print(f'Coefficient of Determination (COD): {cod}')
print(f'R-squared: {r2}')

Mean Absolute Error: 3.0753526709353145e-11
Mean Squared Error: 2.2279737596603523e-21
Coefficient of Determination (COD): 1.0
R-squared: 1.0


### Regularization using Lasso for multicollinearity

In [29]:
from sklearn.linear_model import Lasso

#  L1 regularization (Lasso)
model_lasso = Lasso(alpha=0.01)
model_lasso.fit(X_train, y_train)

# predictions on the testing data
y_pred_lasso = model_lasso.predict(X_test)

# Mean Absolute Error
mae_lasso = mean_absolute_error(y_test, y_pred_lasso)

# Coefficient of Determination (COD)
cod_lasso = r2_score(y_test, y_pred_lasso)

print(f'Mean Absolute Error (Lasso): {mae_lasso}')
print(f'Coefficient of Determination (COD) - Lasso: {cod_lasso}')


Mean Absolute Error (Lasso): 5.345919180242254
Coefficient of Determination (COD) - Lasso: 0.9999998454022021


###### regularization has not significantly affected the performance of your model.

In [30]:
lasso_coefficients = pd.DataFrame({'Feature': X.columns, 'Coefficient': model_lasso.coef_})
print(lasso_coefficients)

                    Feature   Coefficient
0                Unnamed: 0      0.916882
1    Combustible Renewables     50.139208
2                     Hydro  18996.894010
3          Other Renewables      1.475723
4                     Solar   4168.727271
..                      ...           ...
219           Country_Spain    -23.830017
220          Country_Sweden    -13.424351
221     Country_Switzerland    -12.214560
222  Country_United Kingdom    -28.308225
223   Country_United States     35.591415

[224 rows x 2 columns]


### One Hot enconder not scaled

In [31]:
#categorical columns
categorical_columns = df.select_dtypes(include=['object']).columns

#one-hot encoding to categorical columns
X = pd.get_dummies(df, columns=categorical_columns, drop_first=True)
X

Unnamed: 0  Combustible Renewables       Hydro  Other Renewables  \
0              0                216.2870   1044.4060            0.0000   
1              1                350.3830   2504.1300            0.0000   
2              2                383.1770    141.8980            0.0000   
3              3                712.9880  28243.7380            2.4400   
4              4                152.8260   1740.4160            0.0000   
...          ...                     ...         ...               ...   
6713        6713                478.4796   1459.6225            1.5000   
6714        6714                819.7182   5992.2339            0.0000   
6715        6715                173.8267   2753.3512            0.0000   
6716        6716               3122.6177    367.4721            0.4559   
6717        6717               4322.0950  18246.4312            0.0000   

           Solar  Total Renewables (Hydro, Geo, Solar, Wind, Other)  \
0        26.8110                                          1638.0980   
1         9.9650                                          2995.6960   
2        80.4190                                           702.5090   
3        24.1840                                         29709.3010   
4         0.0000                                          1913.6010   
...          ...                                                ...   
6713   3213.8391                                          9351.9487   
6714    101.3747                                          8900.2441   
6715    337.7122                                          3276.0098   
6716   1385.7783                                         10397.5687   
6717  17437.4229                                         68452.5391   

            Wind  value_CDD_18  value_CDD_21  \
0       350.5110      48.13000     12.620000   
1       131.1070       0.05449      0.000000   
2        97.0150       0.99160      0.000000   
3       725.9510       0.06883      0.000023   
4        20.3590       1.41100      0.040190   
...          ...           ...           ...   
6713   4198.5075     112.80000     48.410000   
6714   1986.9173       0.00000      0.000000   
6715     11.1196       3.99700      0.000000   
6716   5521.2446       2.85300      0.000000   
6717  27065.9698     137.80000     74.390000   

      value_Global_Horizontal_Irrandiance  ...  Country_Portugal  \
0                                720994.0  ...             False   
1                                736161.0  ...             False   
2                                723195.0  ...             False   
3                                679927.0  ...             False   
4                                578355.0  ...             False   
...                                   ...  ...               ...   
6713                             748831.0  ...             False   
6714                             314577.0  ...             False   
6715                             565099.0  ...             False   
6716                             422664.0  ...             False   
6717                             697004.0  ...             False   

      Country_Romania  Country_Serbia  Country_Slovak Republic  \
0               False           False                    False   
1               False           False                    False   
2               False           False                    False   
3               False           False                    False   
4               False           False                    False   
...               ...             ...                      ...   
6713            False           False                    False   
6714            False           False                    False   
6715            False           False                    False   
6716            False           False                    False   
6717            False           False                    False   

      Country_Slovenia  Country_Spain  Country_Sweden  Country_Switzerland  \

### One Hot enconder scaled

In [90]:
import pandas as pd
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, OneHotEncoder

df = pd.read_csv('/Users/bryan/code/VonRiecken/Power-Predict/power_predict/data/merged_dataset.csv')

categorical_columns = df.select_dtypes(include=['object']).columns

numerical_columns = df.select_dtypes(exclude=['object']).columns

# transformers for one-hot encoding and scaling
preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numerical_columns),
        ('cat', OneHotEncoder(drop='first', sparse=False), categorical_columns)
    ])

# pipeline
pipeline = Pipeline(steps=[('preprocessor', preprocessor)])

# Fit
df_transformed = pipeline.fit_transform(df)

# feature names after one-hot encoding
feature_names = list(numerical_columns) + list(pipeline.named_steps['preprocessor'].named_transformers_['cat'].get_feature_names_out(categorical_columns))

# Convert the transformed array back to a DataFrame
df_transformed = pd.DataFrame(df_transformed, columns=feature_names)

df_transformed.head()


/Users/bryan/.pyenv/versions/3.10.6/envs/Power-Predict/lib/python3.10/site-packages/sklearn/preprocessing/_encoders.py:975: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


Unnamed: 0  Combustible Renewables     Hydro  Other Renewables     Solar  \
0   -1.731793               -0.371839 -0.288311         -0.219984 -0.279923   
1   -1.731277               -0.284767 -0.186769         -0.219984 -0.285258   
2   -1.730762               -0.263473 -0.351091         -0.219984 -0.262948   
3   -1.730246               -0.049320  1.603731          0.079841 -0.280755   
4   -1.729730               -0.413045 -0.239895         -0.219984 -0.288413   

   Total Renewables (Hydro, Geo, Solar, Wind, Other)      Wind  value_CDD_18  \
0                                          -0.316003 -0.268083      0.139958   
1                                          -0.257861 -0.302890     -0.559862   
2                                          -0.356072 -0.308299     -0.546221   
3                                           0.886205 -0.208521     -0.559654   
4                                          -0.304204 -0.320460     -0.540116   

   value_CDD_21  value_Global_Horizontal_Irrandiance  ...  Country_Portugal  \
0     -0.151031                             0.523027  ...               0.0   
1     -0.445247                             0.573405  ...               0.0   
2     -0.445247                             0.530337  ...               0.0   
3     -0.445247                             0.386620  ...               0.0   
4     -0.444310                             0.049241  ...               0.0   

   Country_Romania  Country_Serbia  Country_Slovak Republic  Country_Slovenia  \
0              0.0             0.0                      0.0               0.0   
1              0.0             0.0                      0.0               0.0   
2              0.0             0.0                      0.0               0.0   
3              0.0             0.0                      0.0               0.0   
4              0.0             0.0                      0.0               0.0   

   Country_Spain  Country_Sweden  Country_Switzerland  Country_United Kingdom  \
0            0.0             0.0                  0.0                     0.0   
1            0.0             0.0                  0.0                     0.0   
2            0.0             0.0                  0.0                     0.0   
3            0.0             0.0                  0.0                     0.0   
4            0.0             0.0                  0.0                     0.0   

   Country_United States  
0                    0.0  
1                    0.0  
2                    0.0  
3                    0.0  
4                    0.0  

[5 rows x 224 columns]

### Linear Regression not scaled

In [86]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

X = pd.get_dummies(df, columns=categorical_columns, drop_first=True)
y = df['total target (wind, solar, hydro)']

# Split training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# linear regression model
model = LinearRegression()

# Fit 
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
cod = r2_score(y_test, y_pred)

print(f'Mean Absolute Error: {mae}')
print(f'Mean Squared Error: {mse}')
print(f'Coefficient of Determination (COD): {cod}')
print(f'R-squared: {r2}')

Mean Absolute Error: 8.319501701304857e-10
Mean Squared Error: 9.63137380842184e-19
Coefficient of Determination (COD): 1.0
R-squared: 1.0


### XGBoost unscaled

In [33]:
import xgboost as xgb

X = pd.get_dummies(df, columns=categorical_columns, drop_first=True)
y = df['total target (wind, solar, hydro)']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# XGBoost model
model = xgb.XGBRegressor(objective='reg:squarederror', random_state=42)

# Fit 
model.fit(X_train, y_train)

# predict testing data
y_pred = model.predict(X_test)

# Mean Absolute Error
mae = mean_absolute_error(y_test, y_pred)

# Coefficient of Determination (COD)
cod = r2_score(y_test, y_pred)

print(f'Mean Absolute Error: {mae}')
print(f'Coefficient of Determination (COD): {cod}')


Mean Absolute Error: 177.40084097378679
Coefficient of Determination (COD): 0.9954392355836477


### XGBoost scaled

In [87]:
# numerical features for scaling
numerical_columns = X.select_dtypes(include=['float64', 'int64']).columns

# StandardScaler object
scaler = StandardScaler()

# Scaled numerical features
X[numerical_columns] = scaler.fit_transform(X[numerical_columns])

# One-hot encode categorical columns
X = pd.get_dummies(X, columns=X.select_dtypes(include=['object']).columns, drop_first=True)

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# XGBoost model
model = xgb.XGBRegressor(objective='reg:squarederror', random_state=42)

# Fit the model to the training data
model.fit(X_train, y_train)

# predictions on the testing data
y_pred = model.predict(X_test)

# Mean Absolute Error
mae = mean_absolute_error(y_test, y_pred)

# Coefficient of Determination (COD)
cod = r2_score(y_test, y_pred)

print(f'Mean Absolute Error: {mae}')
print(f'Coefficient of Determination (COD): {cod}')


Mean Absolute Error: 180.5205075637443
Coefficient of Determination (COD): 0.9954820208240055


### Feature importance from the trained XGBoost model


In [35]:
# feature importance from the trained XGBoost model
feature_importance = model.feature_importances_
feature_importance

array([8.02342111e-05, 1.17546401e-03, 5.27775893e-03, 5.09715464e-04,
       9.96511662e-04, 7.86261380e-01, 3.00893706e-04, 6.08995033e-05,
       7.41757802e-04, 2.54378829e-04, 6.15036988e-05, 1.61320713e-05,
       1.15248158e-05, 2.68580887e-04, 6.63963438e-04, 1.93332598e-04,
       1.85044035e-01, 0.00000000e+00, 0.00000000e+00, 1.09752909e-06,
       3.58356124e-06, 6.87443844e-06, 1.34935431e-06, 1.17602326e-06,
       1.52888952e-05, 5.05960361e-06, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 2.56711076e-07, 0.00000000e+00, 3.44644855e-06,
       2.43373370e-06, 2.36383880e-06, 0.00000000e+00, 0.00000000e+00,
       1.76184230e-07, 1.05352308e-06, 0.00000000e+00, 4.88907347e-07,
       1.28903238e-07, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 7.07398442e-07, 0.00000000e+00, 5.73293664e-06,
       1.80070310e-06, 0.00000000e+00, 2.74505624e-07, 0.00000000e+00,
       1.19368178e-06, 0.00000000e+00, 2.20957577e-06, 0.00000000e+00,
      

### 

####

### Pipeline

### Train-test split to assess the generalization performance of your model

In [42]:
from sklearn.model_selection import train_test_split

X = df.drop("total target (wind, solar, hydro)", axis=1)
y = df["total target (wind, solar, hydro)"]

random_state = 42

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=random_state)


### Linear Regression with feature scaling

In [44]:
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error

# Exclude non-numeric columns from scaling
numeric_columns = X_train.select_dtypes(include=['number']).columns
X_train_numeric = X_train[numeric_columns]
X_test_numeric = X_test[numeric_columns]

# feature scaling using StandardScaler
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train_numeric)
X_test_scaled = scaler.transform(X_test_numeric)

# Initialize and train the Linear Regression model
linear_reg_model = LinearRegression()
linear_reg_model.fit(X_train_scaled, y_train)

#predictions on the test set
y_pred = linear_reg_model.predict(X_test_scaled)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
print(f'Mean Squared Error: {mse}')


Mean Squared Error: 4.825849893735199e-22


### Linear Regression with default parameters

In [48]:
linear_reg_default_model = LinearRegression()
linear_reg_default_model.fit(X_train_numeric, y_train)

y_pred_default = linear_reg_default_model.predict(X_test_numeric)

mse_default = mean_squared_error(y_test, y_pred_default)
print(f'Mean Squared Error (Default Parameters): {mse_default}')


Mean Squared Error (Default Parameters): 1.4968430922743412e-21


 ### Pipeline Initialization

In [55]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, OneHotEncoder

# Identify numeric and non-numeric columns
numeric_features = X_train.select_dtypes(include=['number']).columns
non_numeric_features = X_train.select_dtypes(exclude=['number']).columns

# Create transformers
numeric_transformer = Pipeline(steps=[
    ('scaler', StandardScaler())
])

non_numeric_transformer = Pipeline(steps=[
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

# ColumnTransformer to apply different preprocessing to numeric and non-numeric columns
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('non_num', non_numeric_transformer, non_numeric_features)
    ])

pipeline = Pipeline([
    ('preprocessor', preprocessor),
    ('regression', LinearRegression())
])

pipeline.fit(X_train, y_train)

y_pred = pipeline.predict(X_test)
y_pred

array([1.06452861e+02, 5.40937336e+04, 1.18377948e+02, ...,
       9.25151013e+02, 5.08221437e+01, 6.60015586e+03])

### Hyperparameter Tuning using GridSearchCV

In [74]:
print(df.columns)

numeric_features = df.select_dtypes(include=['float64', 'int64']).columns.tolist()

target_column_name = 'total target (wind, solar, hydro)'
if target_column_name in numeric_features:
    numeric_features.remove(target_column_name)

if 'Month_year' in numeric_features:
    numeric_features.remove('Month_year')
else:
    categorical_features.append('Month_year')

print("Numeric Features:", numeric_features)
print("Categorical Features:", categorical_features)


Index(['Unnamed: 0', 'Month_year', 'Country', 'Balance',
       'Combustible Renewables', 'Hydro', 'Other Renewables', 'Solar',
       'Total Renewables (Hydro, Geo, Solar, Wind, Other)', 'Wind',
       'value_CDD_18', 'value_CDD_21', 'value_Global_Horizontal_Irrandiance',
       'value_HDD_16', 'value_HDD_18', 'value_Heat_index',
       'value_Relative_Humidty', 'value_Temperature',
       'value_Total_Precipitation', 'total target (wind, solar, hydro)'],
      dtype='object')
Numeric Features: ['Unnamed: 0', 'Combustible Renewables', 'Hydro', 'Other Renewables', 'Solar', 'Total Renewables (Hydro, Geo, Solar, Wind, Other)', 'Wind', 'value_CDD_18', 'value_CDD_21', 'value_Global_Horizontal_Irrandiance', 'value_HDD_16', 'value_HDD_18', 'value_Heat_index', 'value_Relative_Humidty', 'value_Temperature', 'value_Total_Precipitation']
Categorical Features: ['Month_year', 'Country', 'Balance', 'Month_year', 'Month_year', 'Month_year']


### Pipeline with preprocessing

In [83]:
from sklearn.model_selection import train_test_split, RandomizedSearchCV
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestRegressor
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder

X = df.drop(target_column_name, axis=1)
y = df[target_column_name]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),
    ('scaler', StandardScaler())
])

categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='missing')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)
    ])

pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                             ('regressor', RandomForestRegressor())])

param_grid = {
    'regressor__n_estimators': [50, 100, 200],
    'regressor__max_depth': [None, 10, 20, 30],
    'regressor__min_samples_split': [2, 5, 10],
    'regressor__min_samples_leaf': [1, 2, 4]
}

random_search = RandomizedSearchCV(pipeline, param_distributions=param_grid, n_iter=10,
                                   cv=5, scoring='neg_mean_squared_error', n_jobs=-1, random_state=42)

random_search.fit(X_train, y_train)

print("Best Hyperparameters:", random_search.best_params_)


Best Hyperparameters: {'regressor__n_estimators': 200, 'regressor__min_samples_split': 2, 'regressor__min_samples_leaf': 2, 'regressor__max_depth': None}


###### from above, number of trees = 200 (A higher number of trees may lead to better model performance, but it also increases computation time.)

###### regressor__min_samples_split(minimum number of samples required to split an internal node). Smaller value (2) means that the model is more sensitive to noise, possibly overfitting the training data.

###### regressor__min_samples_leaf(minimum number of samples required to be at a leaf node) Smaller value (2) allows the model to make more complex decisions but might again lead to overfitting.

###### regressor__max_depth (maximum depth of the individual trees). A None value means that nodes will expand until they contain less than min_samples_split samples. A deeper tree can capture more complex patterns, but is more prone to overfitting.

### Model Evaluation: Cross-Validation

In [84]:
from sklearn.model_selection import cross_val_score, KFold

num_folds = 5 
kf = KFold(n_splits=num_folds, shuffle=True, random_state=42)

# cross-validation
cv_scores = cross_val_score(random_search.best_estimator_, X_train, y_train, cv=kf, scoring='neg_mean_squared_error', n_jobs=-1)

# mean and standard deviation of the cross-validation scores
mean_cv_score = np.mean(cv_scores)
std_cv_score = np.std(cv_scores)

print(f"Cross-Validation Mean Squared Error: {mean_cv_score}")
print(f"Cross-Validation Standard Deviation: {std_cv_score}")


Cross-Validation Mean Squared Error: -327137.61235716665
Cross-Validation Standard Deviation: 185082.44589041328


######  MSE=Value: -327137.61235716665  negative scoring function used (neg_mean_squared_error in scikit-learn). Lower values are better, and the closer to zero, the better the model's performance.

###### CVScore=185082.44589041328 high standard deviation indicates some variability in the model's performance across folds.

### Fit the best model on the entire training dataset


In [ ]:

# final_model = random_search.best_estimator_
# final_model.fit(X_train, y_train)



### Evaluate the final model on the test set.

In [ ]:
# from sklearn.metrics import mean_squared_error

# # predictions on test set
# y_pred = final_model.predict(X_test)

# # Mean squared error
# mse = mean_squared_error(y_test, y_pred)

# print("Mean Squared Error on Test Set:", mse)
